### DISCUSSION TITLE
Issue while creating production build

### DISCUSSION THREAD

**Discussion by Yaswanth-Adeption** [vercel/next.js#60692] [App Router] [answered]
I am using Next js v14.0.4 (app router) and React v18.2.0, and 
this is my **package json**
```
{
  "name": "adeption-website-next-upgrade",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "lint": "next lint",
    "analyze": "cross-env ANALYZE=true next build",
    "analyze:server": "cross-env BUNDLE_ANALYZE=server next build",
    "analyze:browser": "cross-env BUNDLE_ANALYZE=browser next build",
    "lint:fix": "eslint --fix",
    "format": "prettier --write . --config ./.prettierrc.json",
    "prepare": "husky install"
  },
  "dependencies": {
    "@azure/storage-blob": "^12.17.0",
    "@commitlint/cli": "^18.4.4",
    "@commitlint/config-conventional": "^18.4.4",
    "@emotion/cache": "^11.11.0",
    "@emotion/react": "^11.11.3",
    "@emotion/styled": "^11.11.0",
    "@mui/icons-material": "^5.15.4",
    "@mui/lab": "^5.0.0-alpha.158",
    "@mui/material": "^5.15.4",
    "@mui/system": "^5.15.4",
    "@mui/x-date-pickers": "^6.19.0",
    "@next/bundle-analyzer": "^14.0.4",
    "@reduxjs/toolkit": "^2.0.1",
    "@stripe/react-stripe-js": "^2.4.0",
    "@stripe/stripe-js": "^2.3.0",
    "@tinymce/tinymce-react": "^4.3.2",
    "axios": "^1.6.5",
    "chart.js": "^4.4.1",
    "chartjs-plugin-datalabels": "^2.2.0",
    "cookies-next": "^4.1.0",
    "crypto-js": "^4.2.0",
    "dayjs": "^1.11.10",
    "emoji-picker-element": "^1.21.0",
    "eslint-plugin-import": "^2.29.1",
    "eslint-plugin-node": "^11.1.0",
    "eslint-plugin-prettier": "^5.1.3",
    "eslint-plugin-promise": "^6.1.1",
    "html2canvas": "^1.4.1",
    "husky": "^8.0.3",
    "jspdf": "^2.5.1",
    "moment-timezone": "^0.5.44",
    "next": "^14.0.4",
    "next-auth": "^4.24.5",
    "next-redux-wrapper": "^8.1.0",
    "npm": "^10.3.0",
    "prop-types": "^15.8.1",
    "react": "^18.2.0",
    "react-chartjs-2": "^5.2.0",
    "react-country-flag": "^3.1.0",
    "react-device-detect": "^2.2.3",
    "react-dom": "^18.2.0",
    "react-dom-confetti": "^0.2.0",
    "react-hook-form": "^7.49.3",
    "react-redux": "^9.1.0",
    "react-simply-carousel": "^9.1.1",
    "react-slick": "^0.29.0",
    "sass": "^1.69.7",
    "tag2cloud": "^1.0.6",
    "timeago-react": "^3.0.6"
  },
  "devDependencies": {
    "autoprefixer": "^10.4.16",
    "babel-plugin-transform-imports": "^2.0.0",
    "eslint": "^8.56.0",
    "eslint-config-next": "14.0.3",
    "eslint-config-prettier": "^9.1.0",
    "eslint-config-standard": "^17.1.0",
    "eslint-plugin-react": "^7.33.2",
    "prettier": "3.1.0"
  }
}
```
and this is my **next.config.js**
```
const ContentSecurityPolicy = `...`
const withBundleAnalyzer = require('@next/bundle-analyzer')({
  enabled: process.env.ANALYZE === 'true'
})

module.exports = (phase, { defaultConfig }) => {
  const nextConfig = {
    compiler: {
      styledComponents: true
    },
    env: {
      NEXTAUTH_SECRET: '...'
    },
    images: {
      domains: [
       ...
      ]
    },
    async headers() {
      return [
        {
          source: '/((?!embed).*)',
          headers: [
            {
              key: 'X-Frame-Options',
              value: 'DENY'
            },
            {
              key: 'Referrer-Policy',
              value: 'origin-when-cross-origin'
            },
            {
              key: 'Strict-Transport-Security',
              value: 'max-age=31536000; includeSubDomains; preload'
            },
            {
              key: 'Content-Security-Policy',
              value: ContentSecurityPolicy.replace(/\s{2,}/g, ' ').trim()
            }
          ]
        }
      ]
    }
  }
  return withBundleAnalyzer(nextConfig)
}
```
The project is running fine with `npm run dev` in localhost:3000 but when I am doing next build i.e., `npm run build` I am getting this error 
```
Creating an optimized production build  ...thread '<unnamed>' panicked at C:\Users\Administrator\.cargo\registry\src\index.crates.io-6f17d22bba15001f\swc_ecma_transforms_optimization-0.195.58\src\simplify\expr\mod.rs:624:25:
not implemented: error reporting: FRACTIONAL_BITWISE_OPERAND
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
![image](https://github.com/vercel/next.js/assets/156776457/f00e2e63-c0c2-4ba0-a353-70777851034e)
 
Tried running with RUST_BACKTRACE=1 and RUST_BACKTRACE=full
and there is no use.

Can anyone please help on how to proceed further or any solution for this
Please feel free to ask for further information

**Comment by icyJoseph** [comment, marked as answer]
Hi,

Alright, this seems to be an issue with the SWC compiler, but I cannot really tell where.

You could try disabling the SWC minify, even for the server:

- https://github.com/vercel/next.js/discussions/58182#discussioncomment-7514351

```js
//next.config.js
module.exports = {
    experimental: {
        serverMinification: false,
        ....
    },
   swcMinify: false,
    ....
}
```

Does that help it build?

**Comment by Yaswanth-Adeption** [comment]
Hey @icyJoseph, thank you so much, that change helped with building the Application but I have a small doubt regarding this as its showing an caution message while using disabling the SWC Minify on both server and client side
```
Disabling SWC Minifer will not be an option in the next major version.
```
![image](https://github.com/vercel/next.js/assets/156776457/343daa72-38ef-4845-abb0-899fbd8d91c1)

My doubt is that, will this be solved in the next major release as there wont be an option available to disable swcMinify otherwise do we need to any changes in order to prevent these errors

**Reply by icyJoseph** [reply]
Yes, so we got step 1 covered. We can now be certain that this is an SWC error. 

Now, as the years have gone by, SWC has grown and covered a lot of edge cases regarding JavaScript compilation. 

So every time a new issue comes up it gets even tricker to narrow down the root issue. 

Which minify fixes things? experimental.serverMinification, or swcMinify? If it is the experimental flag, then it means that some server side code of yours is causing issues. If it is the `swcMinify` flag, then it'd have to be a browser code problem.

Let's try to narrow down which one is it.

**Reply by Yaswanth-Adeption** [reply]
I have actually tried disabling swcMinify flag and that gave me an issue 
```
Failed to compile.

8816.js from Terser
Unexpected token: punc ({) [8816.js:586,10]
    at js_error (E:\Adeption-website\adeption-website-next-upgrade\node_modules\next\dist\compiled\terser\bundle.min.js:1:32780)
    at croak (E:\Adeption-website\adeption-website-next-upgrade\node_modules\next\dist\compiled\terser\bundle.min.js:1:43882)
    at token_error (E:\Adeption-website\adeption-website-next-upgrade\node_modules\next\dist\compiled\terser\bundle.min.js:1:43989)
    at unexpected (E:\Adeption-website\adeption-website-next-upgrade\node_modules\next\dist\compiled\terser\bundle.min.js:1:44055)
    at as_property_name (E:\Adeption-website\adeption-website-next-upgrade\node_modules\next\dist\compiled\terser\bundle.min.js:1:68487)
    at concise_method_or_getset (E:\Adeption-website\adeption-website-next-upgrade\node_modules\next\dist\compiled\terser\bundle.min.js:1:64746)
    at class_ (E:\Adeption-website\adeption-website-next-upgrade\node_modules\next\dist\compiled\terser\bundle.min.js:1:64098)
    at statement (E:\Adeption-website\adeption-website-next-upgrade\node_modules\next\dist\compiled\terser\bundle.min.js:1:46908)
    at _embed_tokens_wrapper (E:\Adeption-website\adeption-website-next-upgrade\node_modules\next\dist\compiled\terser\bundle.min.js:1:45013)
    at block_ (E:\Adeption-website\adeption-website-next-upgrade\node_modules\next\dist\compiled\terser\bundle.min.js:1:56881)


> Build failed because of webpack errors
   Creating an optimized production build  .
```
And if I only add experimental tag then it gave me this issue
```
Creating an optimized production build  ...thread '<unnamed>' panicked at C:\Users\Administrator\.cargo\registry\src\index.crates.io-6f17d22bba15001f\swc_ecma_transforms_optimization-0.195.58\src\simplify\expr\mod.rs:624:25:
not implemented: error reporting: FRACTIONAL_BITWISE_OPERAND
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

And if I add experimental tag along with swcMinify flag then the build getting created successfully.
Does that mean, my code is faulty? 
if yes, then how can I make sure to correct it, as there is no enough information available in those error messages

**Reply by icyJoseph** [reply]
Oh, so you need both, darn...

It is not necessarily that your code is faulty. I think it is more likely, that a Babel code transformation, was either too loose, or not well defined back when it was written, and now that we use SWC to transform code, the assumptions might have changed.

To be precise, either your code, or third party (node_modules, etc), has some JavaScript expression that Babel did understand, but SWC can't. 

`FRACTIONAL_BITWISE_OPERAND` is our hint...

Why did you have, `babel-plugin-transform-imports`?

**Reply by icyJoseph** [reply]
I went to the SWC repository and found this bit:

- https://github.com/swc-project/swc/blob/0763e7e6d60ba15eacc3228bcde20879f465a591/crates/swc_ecma_transforms_optimization/src/simplify/expr/mod.rs#L618

So, it is not implemented at all 🤔 ~ and I wonder, if you have somewhere a bit shift that's using floats?

**Reply by icyJoseph** [reply]
Here you go, this is a minimal reproduction repository, that's not very useful at all, because the code branch is not implemented... but it shows you what the error is (open the console and you'll see `FRACTIONAL_BITWISE_OPERAND`):

- [SWC Playground](https://play.swc.rs/?version=1.3.100&code=H4sIAAAAAAAAAzNUsLFRMNQzBQDRyr4sCAAAAA%3D%3D&config=H4sIAAAAAAAAA42VS47bMAxA93OKwOsu2kWLogformcQFItylMqiIVKZBIPcvbRsZzJj2ugmiPn4ESmSens5HJoztc2vw5v8lY%2FBZoL8%2BBYJ3RLbq0gaaHtLbQ4DN18WeqYReRsJqug%2BkYZt7oCrFX2f1ZuISLCoz7I%2BpOBvzwFb7IcMRE8ykYq%2F0kNi%2Bmg%2Fs4yvI%2BBcnuVHxAg27RBjyYTE0EHWHLcYox0IzMVmxct4UpsDoRZihIXBmSHjoPLkAgdMEnNNHVhnWnSgoJCh5XABzUxiiVkiSU%2FJp2IHx9J19ZI%2FWcPFxmJZiQnXeiVyWsXrCQOx8SVpJZzgRg0mOBf3s2XwJgOXnNZ2Zwxp407%2BAkgFoiVKtgfNb9Xw0k9b1n7XMiQvLcs3hUt%2Fa1km6KSoJgSvVHasDGQO2m1mcKWFsbKtdpwZb5SPggMD3kuvKK7pNXB70oLybQD0CpD7tV7rqgmYxxRu8HEgdvBvyZL1Bps1esunbUq3%2FohxJ0APfEK3oyBXwbiNs2yJ67DNS3IgrQFOVSlUwXoJyAAwmliX5ao3ZDzEo%2BkiHt%2FXxKxwfyzh3qauzvvTvmQcIlwgbrXxf4zILpWzXcauXiZ7Nb3wU2tKK4MXvn398GZIKi%2FLb02q6dGVp4RqT05vyY%2FmXWl5OZaqNYH%2BLIa1Tvd%2FDWFcTeAGAAA%3D)

Do you recognize such a pattern in your code base?

**Reply by Yaswanth-Adeption** [reply]
Hey first of all thank you for your patience and also pointing out the issue.

I have searched all my code and I haven't found any such pattern exists in the code base.

I am not sure if its indicating any other thing.

And regarding `babel-plugin-transform-imports` it is from our old code base, I haven't added it previously in this code base but it gave error and I thought may be this would be the reason for that and added it. Even I removed it now and tried with experimental tag and its still giving the same BITWISE OPERAND error

**Reply by Yaswanth-Adeption** [reply]
To be more precise we haven't used any bit shifting at all in the code base.
So I am thinking that maybe the Third Party libraries that we are using may have that expression used in their code, but how can we find that out?
Would it be done by removing the third party packages one by one to filter out or there is any other way to do so...

**Reply by icyJoseph** [reply]
uff yeah, right now I think your best call is to remove entire pages, and until you find the one, or ones, that are crashing the process, and figure out which packages within those are culprit.. :/ and then you'd need to find a replacement, or make a PR to that package with a fix.

**Reply by Yaswanth-Adeption** [reply]
Oh man :( okay will do that and one more thing,
Regarding the issue which I got by doing `swcMinify: false` which is 
```
Failed to compile.

8816.js from Terser
Unexpected token: punc ({) [8816.js:586,10]
    at js_error (E:\Adeption-website\adeption-website-next-upgrade\node_modules\next\dist\compiled\terser\bundle.min.js:1:32780)
    at croak (E:\Adeption-website\adeption-website-next-upgrade\node_modules\next\dist\compiled\terser\bundle.min.js:1:43882)
    at token_error (E:\Adeption-website\adeption-website-next-upgrade\node_modules\next\dist\compiled\terser\bundle.min.js:1:43989)
    at unexpected (E:\Adeption-website\adeption-website-next-upgrade\node_modules\next\dist\compiled\terser\bundle.min.js:1:44055)
    at as_property_name (E:\Adeption-website\adeption-website-next-upgrade\node_modules\next\dist\compiled\terser\bundle.min.js:1:68487)
    at concise_method_or_getset (E:\Adeption-website\adeption-website-next-upgrade\node_modules\next\dist\compiled\terser\bundle.min.js:1:64746)
    at class_ (E:\Adeption-website\adeption-website-next-upgrade\node_modules\next\dist\compiled\terser\bundle.min.js:1:64098)
    at statement (E:\Adeption-website\adeption-website-next-upgrade\node_modules\next\dist\compiled\terser\bundle.min.js:1:46908)
    at _embed_tokens_wrapper (E:\Adeption-website\adeption-website-next-upgrade\node_modules\next\dist\compiled\terser\bundle.min.js:1:45013)
    at block_ (E:\Adeption-website\adeption-website-next-upgrade\node_modules\next\dist\compiled\terser\bundle.min.js:1:56881)


> Build failed because of webpack errors
   Creating an optimized production build  .
```
I saw the similar issue stated in one of the issues, link to the same [https://github.com/vercel/next.js/issues/59046](url)

It's closed saying it is fixed in Canary, so may I expect that it wont be happening from the next release.
And will that resolve the issue I am facing as its same thing.
